lib/fizzy/utf8.cpp

```diff
@@ -0,0 +1,153 @@
+#include "utf8.hpp"
+#include "types.hpp"
+
+/*
+ * The Unicode Standard, Version 6.0
+ * (https://www.unicode.org/versions/Unicode6.0.0/ch03.pdf)
+ *
+ * Page 94, Table 3-7. Well-Formed UTF-8 Byte Sequences
+ *
+ * +--------------------+------------+-------------+------------+-------------+
+ * | Code Points        | First Byte | Second Byte | Third Byte | Fourth Byte |
+ * +--------------------+------------+-------------+------------+-------------+
+ * | U+0000..U+007F     | 00..7F     |             |            |             |
+ * +--------------------+------------+-------------+------------+-------------+
+ * | U+0080..U+07FF     | C2..DF     | 80..BF      |            |             |
+ * +--------------------+------------+-------------+------------+-------------+
+ * | U+0800..U+0FFF     | E0         | A0..BF      | 80..BF     |             |
+ * +--------------------+------------+-------------+------------+-------------+
+ * | U+1000..U+CFFF     | E1..EC     | 80..BF      | 80..BF     |             |
+ * +--------------------+------------+-------------+------------+-------------+
+ * | U+D000..U+D7FF     | ED         | 80..9F      | 80..BF     |             |
+ * +--------------------+------------+-------------+------------+-------------+
+ * | U+E000..U+FFFF     | EE..EF     | 80..BF      | 80..BF     |             |
+ * +--------------------+------------+-------------+------------+-------------+
+ * | U+10000..U+3FFFF   | F0         | 90..BF      | 80..BF     | 80..BF      |
+ * +--------------------+------------+-------------+------------+-------------+
+ * | U+40000..U+FFFFF   | F1..F3     | 80..BF      | 80..BF     | 80..BF      |
+ * +--------------------+------------+-------------+------------+-------------+
+ * | U+100000..U+10FFFF | F4         | 80..8F      | 80..BF     | 80..BF      |
+ * +--------------------+------------+-------------+------------+-------------+
+ */
+
+// Ranges for 2nd byte
+enum class Rule : unsigned
+{
+    Range80BF,  // 80..BF
```

Comment:
Could also rename the enum to `Range` and drop it from these, e.g. `Range::80BF`.

Reply:
Sounds good to rename, but I don't think identifier starting with a number (`80BF`) is possible

Reply:
You cannot use identifier `80BF`.

Reply:
Yeah realised that after.

Reply:
So should we rename this to something?

Reply:
I'm think I'm fine with current names, too.